### PR TITLE
feat(ADR-024): Enable BTC-timestamping public randomness

### LINKED ISSUE #8: BTC-timestamping public randomness
In the Babylon Bitcoin-staking protocol, finality providers are required to commit a range of public randomness before sending finality signatures. However, if the honest majority assumption does not hold, the attacker might commit conflicting randomness on forks and escape from being slashed with performing equivocation attack. Therefore, BTC-timestamping is introduced to address the issue.

Required changes from Babylon:
* Introduce epoch to each randomness commit to enforce that the committed randomness takes effect only after the relevant epoch is finalized
* Do not assign voting power to finality providers that do not have finalized randomness for a particular block

### PR DESCRIPTION
**Pull request by gitferry**
This PR collects sub-PRs to close #8 and complete the whole feature defined in [ADR-024](https://github.com/babylonlabs-io/pm/blob/main/adr/adr-024-timestamping-public-randomness.md)

TODOs
- [x] #12 
- [x] #23 
- [x] [changes in the finality provider repo](https://github.com/babylonlabs-io/finality-provider/pull/40)
- [x] #47 
- [x] [table-driven tests on VotingPowerDistCache](https://github.com/babylonlabs-io/babylon/pull/38)
- [x] [validate upgrade with db schema change in local deployment](https://github.com/babylonlabs-io/babylon-deployment/tree/gai/upgrade-pub-rand-commit)

### CODE REVIEW

File: proto/babylon/btcstaking/v1/incentive.proto [hunk 1]
```diff
@@ -9,9 +9,15 @@ option go_package = "github.com/babylonlabs-io/babylon/x/btcstaking/types";
 // VotingPowerDistCache is the cache for voting power distribution of finality providers
 // and their BTC delegations at a height
 message VotingPowerDistCache {
+    option (gogoproto.goproto_getters) = false;
+    // total_voting_power is the total voting power of all the finality providers
+    // in the cache
     uint64 total_voting_power = 1;
     // finality_providers is a list of finality providers' voting power information
     repeated FinalityProviderDistInfo finality_providers = 2;
+    // num_active_fps is the number of finality providers that have positive voting power
```
Comment (SebastianElvis) [inline proto/babylon/btcstaking/v1/incentive.proto:18]:
```suggestion
    // num_active_fps is the number of finality providers that have active BTC delegations
```

a FP has positive voting power at a height only when it has active BTC
delegations at this height AND has timestamped public randomness at this height

File: test/e2e/configurer/chain/queries.go [hunk 2]
```diff
@@ -180,7 +180,9 @@ func (n *NodeConfig) QueryListSnapshots() ([]*cmtabcitypes.Snapshot, error) {
 func (n *NodeConfig) QueryRawCheckpoint(epoch uint64) (*ct.RawCheckpointWithMetaResponse, error) {
 	path := fmt.Sprintf("babylon/checkpointing/v1/raw_checkpoint/%d", epoch)
 	bz, err := n.QueryGRPCGateway(path, url.Values{})
-	require.NoError(n.t, err)
+	if err != nil {
+		return nil, err
+	}
```
Comment (SebastianElvis) [inline test/e2e/configurer/chain/queries.go:185]:
```suggestion
	require.NoError(n.t, err)
```
Failing to query the node should not happen so we can panick here

File: test/e2e/configurer/chain/queries_btcstaking.go [hunk 3]
```diff
@@ -94,15 +94,19 @@ func (n *NodeConfig) QueryUnbondedDelegations() []*bstypes.BTCDelegationResponse
...
+	if err != nil {
+		return 0, err
+	}
 
 	var resp bstypes.QueryActivatedHeightResponse
 	err = util.Cdc.UnmarshalJSON(bz, &resp)
-	require.NoError(n.t, err)
+	if err != nil {
+		return 0, err
+	}
```
Comment (SebastianElvis) [inline test/e2e/configurer/chain/queries_btcstaking.go:107]: We can panick here as this means there is something fishy in the setup

**Review by SebastianElvis — APPROVED** [review]
Great work putting everything together!

Only super minor comments but these can be done in subsequent PRs

**Review by KonradStaniec — APPROVED** [review]
LGTM!

File: proto/babylon/btcstaking/v1/incentive.proto (hunk 1 above)
Comment (gitferry) [reply proto/babylon/btcstaking/v1/incentive.proto:18]: Agreed. The two conditions combined lead to fp having voting power

**Review by RafilxTenfen — APPROVED** [review]
LGTM

**Pull request merged**